pkg/dyninst/decode/marshal.go

```diff
@@ -0,0 +1,473 @@
+}
+
+type probeData struct {
+	ID       string       `json:"id,omitempty"`
+	Location locationData `json:"location"`
+}
+
+type locationData struct {
+	Method string `json:"method,omitempty"`
+	File   string `json:"file,omitempty"`
+	Line   int    `json:"line,omitzero"`
+	Type   string `json:"type,omitempty"`
+}
+
+type captureData struct {
+	Entry capturePointData `json:"entry"`
+}
+
+type capturePointData struct {
+	Arguments argumentsData `json:"arguments"`
+}
+
+type argumentsData struct {
+	rootData              []byte                          `json:"-"`
+	rootType              *ir.EventRootType               `json:"-"`
+	addressReferenceCount map[typeAndAddr]output.DataItem `json:"-"`
+	event                 Event                           `json:"-"`
+	decoder               *Decoder                        `json:"-"`
+}
+
+func (ad *argumentsData) MarshalJSONTo(enc *jsontext.Encoder) error {
+	var err error
+	currentlyEncoding := map[typeAndAddr]struct{}{}
+
+	if err = writeTokens(enc,
+		jsontext.BeginObject,
+	); err != nil {
+		return err
+	}
+
+	// We iterate over the 'Expressions' of the EventRoot which contains
+	// metadata and raw bytes of the parameters of this function.
+	for _, expr := range ad.rootType.Expressions {
+		parameterType := expr.Expression.Type
+		parameterData := ad.rootData[expr.Offset : expr.Offset+parameterType.GetByteSize()]
+		if err = writeTokens(enc,
+			jsontext.String(expr.Name),
+			jsontext.BeginObject,
+			jsontext.String("type"),
+			jsontext.String(parameterType.GetName()),
+			jsontext.String("value"),
+		); err != nil {
+			return err
+		}
+		err = ad.decoder.encodeValue(enc, ad.addressReferenceCount, currentlyEncoding, parameterType, parameterData)
+		if err != nil {
+			return fmt.Errorf("error parsing data for field %s: %w", ad.rootType.Name, err)
+		}
+		err = writeTokens(enc, jsontext.EndObject)
+		if err != nil {
+			return err
+		}
+	}
+	if err = writeTokens(enc,
+		jsontext.EndObject,
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+type stackData struct {
+	frames []symbol.StackFrame
+}
+
+func (sd *stackData) MarshalJSONTo(enc *jsontext.Encoder) error {
+	var err error
+	if err = writeTokens(enc, jsontext.BeginArray); err != nil {
+		return err
+	}
+
+	for i := range sd.frames {
+		for j := range sd.frames[i].Lines {
+			if err = json.MarshalEncode(
+				enc, (*stackLine)(&sd.frames[i].Lines[j]),
+			); err != nil {
+				return err
+			}
+		}
+	}
+	if err = writeTokens(enc, jsontext.EndArray); err != nil {
+		return err
+	}
+	return nil
+}
+
+type stackLine gosym.GoLocation
+
+func (sl *stackLine) MarshalJSONTo(enc *jsontext.Encoder) error {
+	if err := writeTokens(enc,
+		jsontext.BeginObject,
+		jsontext.String("function"),
+		jsontext.String(sl.Function),
+		jsontext.String("file"),
+		jsontext.String(sl.File),
+		jsontext.String("line"),
+		jsontext.Int(int64(sl.Line)),
+		jsontext.EndObject,
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *Decoder) encodeValue(
```

Comment:
I'm not certain I see a ton of value in this being a method -- we don't use `d` for much of anything here and I think it can then become something of a false dependency